packages/react-hooks/src/useControllableValue.ts

```diff
@@ -0,0 +1,37 @@
+import * as React from 'react';
+
+/**
+ * Hook to manage a value that could be either controlled or uncontrolled, such as a checked state or
+ * text box string.
+ * @param controlledValue- The controlled value passed in the props. This value will always be used if provided, and the
+ * internal state will be updated to reflect it.
+ * @param defaultUncontrolledValue- Initial value for the internal state in the uncontrolled case.
+ * @see https://reactjs.org/docs/uncontrolled-components.html
+ */
+export function useControllableValue<TValue, TOnChangeArgs extends unknown[]>(
+  controlledValue: TValue | undefined,
+  defaultUncontrolledValue: TValue | undefined,
+  onChange?: (newValue: TValue, ...args: TOnChangeArgs) => void,
+) {
+  const [value, setValue] = React.useState<TValue | undefined>(
+    controlledValue !== undefined ? controlledValue : defaultUncontrolledValue,
+  );
+
+  if (controlledValue !== undefined && controlledValue !== value) {
+    setValue(controlledValue);
+  }
+
+  const setValueOrCallOnChange = React.useCallback(
+    (newValue: TValue, ...args: TOnChangeArgs) => {
+      if (onChange) {
+        onChange(newValue, ...args);
```

Comment:
I'm a bit unsure how this would work. Isn't the first param of `onChange` always an event arg? Why hardcode it to `TValue`?

Reply:
@dzearing Is that true for every `onChange` handler across components? This seems more extensible, and allows for a consistent signature for the case where there is no `onChange` handler.

See #12614 for this hook used in practice.

Reply:
See RFC here: https://github.com/microsoft/fluentui/issues/12588

For sure the intent of `onChange` is to always keep it as predictable as possible. In current code we are using `(ev: FormEvent, newValue: TValue)`, though in the proposed change the second param would be a data object with a predictable way to access `value`, `props` passed to the originating component, and any additional metadata required.